### PR TITLE
Combined amp-img integration for amp-bind

### PR DESCRIPTION
**Pull request by sreenidhipm**
Fix for speed up integration tests for amp-bind (#8197) - Part 1

- Fixes combining test-functions for amp-img integration.
    - Combined separate it() functions into a single function for the amp-img

### CODE REVIEW

**Comment by sreenidhipm** [comment]
I signed it!

File: extensions/amp-bind/0.1/test/test-bind-integration.js [hunk 1]
```diff
@@ -146,62 +146,47 @@ describe.configure().retryOnSaucelabs().run('amp-bind', function() {
     });
   });
 
+  /**
+   * Combined test specs for amp-img
+   */
   describe('amp-img integration', () => {
-    it('should change src when the src attribute binding changes', () => {
-      const button = fixture.doc.getElementById('changeImgSrcButton');
+    it('should do amp-img integration', () => {
```
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: Hmmm, 'should integrate with amp-img' perhaps?

File: extensions/amp-bind/0.1/test/test-bind-integration.js [hunk 2]
```diff
@@ -146,62 +146,47 @@ describe.configure().retryOnSaucelabs().run('amp-bind', function() {
...
+      const changeImgAltButton = fixture.doc
+          .getElementById('changeImgAltButton');
+      const changeImgDimensButton = fixture.doc
+          .getElementById('changeImgDimensButton');
       const img = fixture.doc.getElementById('image');
       expect(img.getAttribute('src')).to.equal('http://www.google.com/image1');
-      button.click();
+      expect(img.getAttribute('height')).to.equal('200');
+      expect(img.getAttribute('width')).to.equal('200');
+      changeImgSrcButton.click();
```
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: I think the goal we're going for here is to try to trigger a change and then verify that it does what we expect. For instance, we expect clicking `changeImgSrcButton` to change the image source, so after it's clicked, we want to verify that it has changed. However, with `invalidSrcButton` and `ftpSrcButton` we want to ensure that clicking them doesn't actually change the binding of src.

What you can do here is click a button, wait for bind to apply the changes, use `expect` to see if the src is correct and then repeat for each one.

File: extensions/amp-bind/0.1/test/test-bind-integration.js [hunk 3]
```diff
@@ -146,62 +146,47 @@ describe.configure().retryOnSaucelabs().run('amp-bind', function() {
...
+      invalidSrcButton.click();
+      ftpSrcButton.click();
+      changeImgAltButton.click();
```
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: Don't forget to assert that `alt`, `width`, and `height` have changed. :smile:

File: extensions/amp-bind/0.1/test/test-bind-integration.js [hunk 4]
```diff
@@ -146,62 +146,47 @@ describe.configure().retryOnSaucelabs().run('amp-bind', function() {
...
+      changeImgDimensButton.click();
```
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: What would be really cool here would be combining the actions that we do expect to happen into one button. So rather than having three buttons that each change `src`, `alt`, or `width` and `height`, you have one button that changes all three at once. We would still want to keep the other ones `invalidSrcButton` and `ftpSrcButton` distinct since we expect them not to change the `src` attribute.

**Comment by sreenidhipm** [comment]
@kmh287 Yes, I'll work on split off the amp-img integration test into its own files in a separate PR. I just did this one beforehand to ask for if this is what is expected so I could work on the rest. Thank you for reviewing :-)

File: extensions/amp-bind/0.1/test/test-bind-integration.js (hunk 2 above)
Comment (sreenidhipm) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: I am not sure I got what you said. I think there is another expect that checks the image dimensions after the button is clicked. Initially, we expect to have the image in a specific dimension and then the dimension changes after the button is clicked (this is what I understood; please correct me if I'm wrong). 

`expect(img.getAttribute('height')).to.equal('200');
  expect(img.getAttribute('width')).to.equal('200');
  changeImgDimensButton.click();
  // Should change width and height when their bindings change
  expect(img.getAttribute('height')).to.equal('300');
  expect(img.getAttribute('width')).to.equal('300');`

File: extensions/amp-bind/0.1/test/test-bind-integration.js (hunk 4 above)
Comment (sreenidhipm) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: Do I have to combine the setState() for all of these buttons into one in the amp-bind-integrations.html? That way, I guess one button can have all of these functions.

File: extensions/amp-bind/0.1/test/test-bind-integration.js (hunk 2 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: For this I'm referring specifically to the buttons that mutate the image source. `changeImgSrcButton`, `ftpSrcButton`, and `invalidSrcButton` all change the `src` attribute of the image element. In order to test that these buttons do what we expect, we have to test them one by one and then check the result. If all three buttons are clicked at once, we can't know what the outcomes of each of the clicks is, just what the end result is.

File: extensions/amp-bind/0.1/test/test-bind-integration.js (hunk 4 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: Yes exactly :+1:

**Comment by kmh287** [comment]
@sreenidhipm No problem! I was actually thinking about #8197. My understanding is that the task is broken up into two parts. 

1: Merging the test functions for each component into one big test function (`it`)
2: Splitting the integration tests into one file per component

I'm curious which of these two provides the greater speedup. On your next few PRs, how would you feel about just breaking the tests out into their own files and holding off on combining the test functions together?

**Comment by sreenidhipm** [comment]
Sure, will do that asap.

File: extensions/amp-bind/0.1/test/test-bind-integration.js (hunk 2 above)
Comment (sreenidhipm) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: got it!

**Comment by sreenidhipm** [comment]
@kmh287 @choumx : I have done the splitting up amp integration tests, each into a single file under a new folder "amp-bind-integrations" which is in turn under "test/fixtures". Yet to update "test-bind-integrations.js" so that it loads each file per component. Please take a look at the PR submitted. Thanks!

File: test/fixtures/amp-bind-integrations/amp-brightcove.html [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  </style>
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+  <script src="/dist/amp.js"></script>
+  <script custom-element="amp-bind" src="/dist/v0/amp-bind-0.1.max.js"></script>
+  <script custom-element="amp-brightcove" src="/dist/v0/amp-brightcove-0.1.max.js"></script>
+</head>
+
+<body>
+  <p>DYNAMIC TAGS</p>
```
Comment (kmh287) [inline test/fixtures/amp-bind-integrations/amp-brightcove.html]: I think the DYNAMIC TAGS test and the P TAG tests should also be moved to their own files as well. That way we can make each of these test files even more concise :smile:

**Comment by kmh287** [comment]
@sreenidhipm Looking good so far. One thing I would recommend would be spreading the changes for each integration over multiple smaller PRs rather than trying to do them all here in one review i.e. just leave this PR for changes to the img integration test and then make another one for amp-video, amp-carousel etc. That being said, you're making good progress on splitting all the HTML files up. I've left a comment as I think we can make them even more concise. (Though, you will probably want to make the requested changes in the future PRs I mentioned rather than here,)

File: test/fixtures/amp-bind-integrations/amp-brightcove.html (hunk 5 above)
Comment (sreenidhipm) [inline test/fixtures/amp-bind-integrations/amp-brightcove.html]: Done! I have sent separate PRs for those too.

Comment (kmh287) [inline test/fixtures/amp-bind-integrations/amp-brightcove.html]: :+1:

**Comment by kmh287** [comment]
Can you also remove the other HTML files from this PR as they'll be included in the others?

**Comment by sreenidhipm** [comment]
Done! Removed other HTML files (other than the one for amp-img)

**Comment by kmh287** [comment]
@sreenidhipm Thanks!

This PR and all the other ones are coming along nicely! I'm still not quite sure about merging the tests all into one function (I know that's what the issue asks for). My concern is that, while one larger test only needs to load the fixture once and is thus faster, merging multiple discrete tests into one function makes the individual tests more difficult to reason about and makes it harder to understand what's broken when tests fail. If we change something that breaks bind img integration and the test fails 25% of the way through, we won't know if anything tested after the failure in the test function is working until we fix the first failure. 

So what I'm thinking we should do here is wrap up the other PRs you've started that just split the tests and fixtures into different files without merging the test functions and then evaluate if the performance improvement we get from those is enough. Then we can circle back to this PR and decide what to do. How does that sound to you?

I know you're getting conflicting signals from our end. Thank you for understanding.

File: extensions/amp-bind/0.1/test/test-bind-integration.js (hunk 2 above)
Comment (kmh287) [inline extensions/amp-bind/0.1/test/test-bind-integration.js]: Bump (in case this comment got lost)
